Simplify linspace decomp and increase its tolerance

lezcano opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #87182
* #87183
* #87205
* #87204
* __->__ #87203

This is an interesting one

Since this is an operation that's intrinsically defined on the reals,
we should perform the ops on that dtype always, and just cast to
the desired dtype at the end. This simplifies the decomposition.

Now, I started looking at this one when I started seeing failures on a
test that's added in a later PR. What's going on here is that, by doing
an upcast to a higher dtype and then cast down to integers, sometimes
there's an off-by-one error. I think this is fine, as the decomposition
is more accurate than the original function, which goes in line with
the whole PrimTorch effort.

cc @ezyang @mruberry @ngimel @Lezcano @peterbell10 @fdrocha

lezcano committed:
Simplify linspace decomp and increase its tolerance

This is an interesting one

Since this is an operation that's intrinsically defined on the reals,
we should perform the ops on that dtype always, and just cast to
the desired dtype at the end. This simplifies the decomposition.

Now, I started looking at this one when I started seeing failures on a
test that's added in a later PR. What's going on here is that, by doing
an upcast to a higher dtype and then cast down to integers, sometimes
there's an off-by-one error. I think this is fine, as the decomposition
is more accurate than the original function, which goes in line with
the whole PrimTorch effort.

[ghstack-poisoned]

lezcano committed:
Update on "Simplify linspace decomp and increase its tolerance"


This is an interesting one

Since this is an operation that's intrinsically defined on the reals,
we should perform the ops on that dtype always, and just cast to
the desired dtype at the end. This simplifies the decomposition.

Now, I started looking at this one when I started seeing failures on a
test that's added in a later PR. What's going on here is that, by doing
an upcast to a higher dtype and then cast down to integers, sometimes
there's an off-by-one error. I think this is fine, as the decomposition
is more accurate than the original function, which goes in line with
the whole PrimTorch effort.

[ghstack-poisoned]

ngimel commented on torch/_refs/__init__.py:
```diff
@@ -4010,41 +4010,18 @@ def linspace(
-    return ret
+        step_size = (end - start) / (steps - 1)
+        eps = step_size / 2
+        ret_double = torch.arange(  # type: ignore[call-overload]
```
And it doesn't directly answer the question of what to do when start and step are integer (yes yes I know that integer is real, nevertheless, it's usual and customary to explicitly distinguish between them).

lezcano replied:
alas, this op cannot be computed natively (for all I know) in those ops without a casting. Even more, this op is not well-defined on ints. IMO, ints shouldn't be supported on this op, and these things are what we get for supporting them...

lezcano replied:
Now, given the current state of things, I think we get to decide what's the definition, and it can very well be "compute things on real floating precision and then cast down".

lezcano replied:
The current decomposition on double yields (at least locally) the same results than the one that was there before, so I think this change is fine.

I think this is as good as it gets with this decomposition and those annoying inputs really, unless you have a different proposal.

ngimel replied:
as @mruberry suggested above, and per docs, the computation type should be global default type, not double.

lezcano committed:
Update on "Simplify linspace decomp and increase its tolerance"


This is an interesting one

Since this is an operation that's intrinsically defined on the reals,
we should perform the ops on that dtype always, and just cast to
the desired dtype at the end. This simplifies the decomposition.

Now, I started looking at this one when I started seeing failures on a
test that's added in a later PR. What's going on here is that, by doing
an upcast to a higher dtype and then cast down to integers, sometimes
there's an off-by-one error. I think this is fine, as the decomposition
is more accurate than the original function, which goes in line with
the whole PrimTorch effort.

[ghstack-poisoned]

lezcano committed:
Update on "Simplify linspace decomp and increase its tolerance"


This is an interesting one

Since this is an operation that's intrinsically defined on the reals,
we should perform the ops on that dtype always, and just cast to
the desired dtype at the end. This simplifies the decomposition.

Now, I started looking at this one when I started seeing failures on a
test that's added in a later PR. What's going on here is that, by doing
an upcast to a higher dtype and then cast down to integers, sometimes
there's an off-by-one error. I think this is fine, as the decomposition
is more accurate than the original function, which goes in line with
the whole PrimTorch effort.

[ghstack-poisoned]

lezcano committed:
Update on "Simplify linspace decomp and increase its tolerance"


This is an interesting one

Since this is an operation that's intrinsically defined on the reals,
we should perform the ops on that dtype always, and just cast to
the desired dtype at the end. This simplifies the decomposition.

Now, I started looking at this one when I started seeing failures on a
test that's added in a later PR. What's going on here is that, by doing
an upcast to a higher dtype and then cast down to integers, sometimes
there's an off-by-one error. I think this is fine, as the decomposition
is more accurate than the original function, which goes in line with
the whole PrimTorch effort.

[ghstack-poisoned]

lezcano committed:
Update on "Simplify linspace decomp and increase its tolerance"


This is an interesting one

Since this is an operation that's intrinsically defined on the reals,
we should perform the ops on that dtype always, and just cast to
the desired dtype at the end. This simplifies the decomposition.

Now, I started looking at this one when I started seeing failures on a
test that's added in a later PR. What's going on here is that, by doing
an upcast to a higher dtype and then cast down to integers, sometimes
there's an off-by-one error. I think this is fine, as the decomposition
is more accurate than the original function, which goes in line with
the whole PrimTorch effort.

[ghstack-poisoned]

lezcano committed:
Update on "Simplify linspace decomp and increase its tolerance"


This is an interesting one

Since this is an operation that's intrinsically defined on the reals,
we should perform the ops on that dtype always, and just cast to
the desired dtype at the end. This simplifies the decomposition.

Now, I started looking at this one when I started seeing failures on a
test that's added in a later PR. What's going on here is that, by doing
an upcast to a higher dtype and then cast down to integers, sometimes
there's an off-by-one error. I think this is fine, as the decomposition
is more accurate than the original function, which goes in line with
the whole PrimTorch effort.

[ghstack-poisoned]

lezcano committed:
Update on "Simplify linspace decomp and increase its tolerance"


This is an interesting one

Since this is an operation that's intrinsically defined on the reals,
we should perform the ops on that dtype always, and just cast to
the desired dtype at the end. This simplifies the decomposition.

Now, I started looking at this one when I started seeing failures on a
test that's added in a later PR. What's going on here is that, by doing
an upcast to a higher dtype and then cast down to integers, sometimes
there's an off-by-one error. I think this is fine, as the decomposition
is more accurate than the original function, which goes in line with
the whole PrimTorch effort.

[ghstack-poisoned]

lezcano committed:
Update on "Simplify linspace decomp and increase its tolerance"


This is an interesting one

Since this is an operation that's intrinsically defined on the reals,
we should perform the ops on that dtype always, and just cast to
the desired dtype at the end. This simplifies the decomposition.

Now, I started looking at this one when I started seeing failures on a
test that's added in a later PR. What's going on here is that, by doing
an upcast to a higher dtype and then cast down to integers, sometimes
there's an off-by-one error. I think this is fine, as the decomposition
is more accurate than the original function, which goes in line with
the whole PrimTorch effort.

[ghstack-poisoned]

lezcano committed:
Update on "Simplify linspace decomp and increase its tolerance"


This is an interesting one

Since this is an operation that's intrinsically defined on the reals,
we should perform the ops on that dtype always, and just cast to
the desired dtype at the end. This simplifies the decomposition.

Now, I started looking at this one when I started seeing failures on a
test that's added in a later PR. What's going on here is that, by doing
an upcast to a higher dtype and then cast down to integers, sometimes
there's an off-by-one error. I think this is fine, as the decomposition
is more accurate than the original function, which goes in line with
the whole PrimTorch effort.

[ghstack-poisoned]

lezcano committed:
Update on "Simplify linspace decomp and increase its tolerance"


This is an interesting one

Since this is an operation that's intrinsically defined on the reals,
we should perform the ops on that dtype always, and just cast to
the desired dtype at the end. This simplifies the decomposition.

Now, I started looking at this one when I started seeing failures on a
test that's added in a later PR. What's going on here is that, by doing
an upcast to a higher dtype and then cast down to integers, sometimes
there's an off-by-one error. I think this is fine, as the decomposition
is more accurate than the original function, which goes in line with
the whole PrimTorch effort.

[ghstack-poisoned]

lezcano committed:
Update on "Simplify linspace decomp and increase its tolerance"


This is an interesting one

Since this is an operation that's intrinsically defined on the reals,
we should perform the ops on that dtype always, and just cast to
the desired dtype at the end. This simplifies the decomposition.

Now, I started looking at this one when I started seeing failures on a
test that's added in a later PR. What's going on here is that, by doing
an upcast to a higher dtype and then cast down to integers, sometimes
there's an off-by-one error. I think this is fine, as the decomposition
is more accurate than the original function, which goes in line with
the whole PrimTorch effort.

[ghstack-poisoned]

lezcano commented on test/test_decomp.py:
```diff
@@ -200,8 +200,17 @@ def op_assert_equal(test_case, op, test_dtype, orig, decomp, args, kwargs):
         # Exceeds tolerances on CUDA, likely due to fma
         (torch.float32, torch.ops.aten.mv.default) : (1e-5, 3e-5),
         (torch.float64, torch.ops.aten.upsample_bicubic2d.vec) : (1e-5, 1e-6),
+        # The decomposition is TOO correct. It computes everything in int64, so sometimes
+        # there's an off-by-one error. See
+        # https://github.com/pytorch/pytorch/issues/81996
+        # https://github.com/pytorch/pytorch/issues/82230
+        (torch.int8, torch.ops.aten.linspace.default) : (0, 1),
+        (torch.uint8, torch.ops.aten.linspace.default) : (0, 1),
+        (torch.int16, torch.ops.aten.linspace.default) : (0, 1),
+        (torch.int32, torch.ops.aten.linspace.default) : (0, 1),
+        (torch.int64, torch.ops.aten.linspace.default) : (0, 1),
```
We put this one here rather than xfailing the tests, as it works not only for `linspace`, but also for all the functions that use it as a building block.
At any rate, we should figure out what to do with this pesky function.

lezcano committed:
Update on "Simplify linspace decomp and increase its tolerance"


This is an interesting one

Since this is an operation that's intrinsically defined on the reals,
we should perform the ops on that dtype always, and just cast to
the desired dtype at the end. This simplifies the decomposition.

Now, I started looking at this one when I started seeing failures on a
test that's added in a later PR. What's going on here is that, by doing
an upcast to a higher dtype and then cast down to integers, sometimes
there's an off-by-one error. I think this is fine, as the decomposition
is more accurate than the original function, which goes in line with
the whole PrimTorch effort.

[ghstack-poisoned]

lezcano committed:
Update on "Simplify linspace decomp and increase its tolerance"


This is an interesting one

Since this is an operation that's intrinsically defined on the reals,
we should perform the ops on that dtype always, and just cast to
the desired dtype at the end. This simplifies the decomposition.

Now, I started looking at this one when I started seeing failures on a
test that's added in a later PR. What's going on here is that, by doing
an upcast to a higher dtype and then cast down to integers, sometimes
there's an off-by-one error. I think this is fine, as the decomposition
is more accurate than the original function, which goes in line with
the whole PrimTorch effort.

[ghstack-poisoned]

lezcano committed:
Update on "Simplify linspace decomp and increase its tolerance"


This is an interesting one

Since this is an operation that's intrinsically defined on the reals,
we should perform the ops on that dtype always, and just cast to
the desired dtype at the end. This simplifies the decomposition.

Now, I started looking at this one when I started seeing failures on a
test that's added in a later PR. What's going on here is that, by doing
an upcast to a higher dtype and then cast down to integers, sometimes
there's an off-by-one error. I think this is fine, as the decomposition
is more accurate than the original function, which goes in line with
the whole PrimTorch effort.

[ghstack-poisoned]

lezcano committed:
Update on "Simplify linspace decomp and increase its tolerance"


This is an interesting one

Since this is an operation that's intrinsically defined on the reals,
we should perform the ops on that dtype always, and just cast to
the desired dtype at the end. This simplifies the decomposition.

Now, I started looking at this one when I started seeing failures on a
test that's added in a later PR. What's going on here is that, by doing
an upcast to a higher dtype and then cast down to integers, sometimes
there's an off-by-one error. I think this is fine, as the decomposition
is more accurate than the original function, which goes in line with
the whole PrimTorch effort.

[ghstack-poisoned]

lezcano committed:
Update on "Simplify linspace decomp and increase its tolerance"


This is an interesting one

Since this is an operation that's intrinsically defined on the reals,
we should perform the ops on that dtype always, and just cast to
the desired dtype at the end. This simplifies the decomposition.

Now, I started looking at this one when I started seeing failures on a
test that's added in a later PR. What's going on here is that, by doing
an upcast to a higher dtype and then cast down to integers, sometimes
there's an off-by-one error. I think this is fine, as the decomposition
is more accurate than the original function, which goes in line with
the whole PrimTorch effort.

[ghstack-poisoned]

lezcano committed:
Update on "Simplify linspace decomp and increase its tolerance"


This is an interesting one

Since this is an operation that's intrinsically defined on the reals,
we should perform the ops on that dtype always, and just cast to
the desired dtype at the end. This simplifies the decomposition.

Now, I started looking at this one when I started seeing failures on a
test that's added in a later PR. What's going on here is that, by doing
an upcast to a higher dtype and then cast down to integers, sometimes
there's an off-by-one error. I think this is fine, as the decomposition
is more accurate than the original function, which goes in line with
the whole PrimTorch effort.

[ghstack-poisoned]

lezcano committed:
Update on "Simplify linspace decomp and increase its tolerance"


This is an interesting one

Since this is an operation that's intrinsically defined on the reals,
we should perform the ops on that dtype always, and just cast to
the desired dtype at the end. This simplifies the decomposition.

Now, I started looking at this one when I started seeing failures on a
test that's added in a later PR. What's going on here is that, by doing
an upcast to a higher dtype and then cast down to integers, sometimes
there's an off-by-one error. I think this is fine, as the decomposition
is more accurate than the original function, which goes in line with
the whole PrimTorch effort.

cc ezyang mruberry ngimel Lezcano fdrocha

[ghstack-poisoned]

lezcano committed:
Update on "Simplify linspace decomp and increase its tolerance"


This is an interesting one

Since this is an operation that's intrinsically defined on the reals,
we should perform the ops on that dtype always, and just cast to
the desired dtype at the end. This simplifies the decomposition.

Now, I started looking at this one when I started seeing failures on a
test that's added in a later PR. What's going on here is that, by doing
an upcast to a higher dtype and then cast down to integers, sometimes
there's an off-by-one error. I think this is fine, as the decomposition
is more accurate than the original function, which goes in line with
the whole PrimTorch effort.

cc ezyang mruberry ngimel Lezcano fdrocha

[ghstack-poisoned]

lezcano committed:
Update on "Simplify linspace decomp and increase its tolerance"


This is an interesting one

Since this is an operation that's intrinsically defined on the reals,
we should perform the ops on that dtype always, and just cast to
the desired dtype at the end. This simplifies the decomposition.

Now, I started looking at this one when I started seeing failures on a
test that's added in a later PR. What's going on here is that, by doing
an upcast to a higher dtype and then cast down to integers, sometimes
there's an off-by-one error. I think this is fine, as the decomposition
is more accurate than the original function, which goes in line with
the whole PrimTorch effort.

cc ezyang mruberry ngimel Lezcano fdrocha

[ghstack-poisoned]

lezcano committed:
Update on "Simplify linspace decomp and increase its tolerance"


This is an interesting one

Since this is an operation that's intrinsically defined on the reals,
we should perform the ops on that dtype always, and just cast to
the desired dtype at the end. This simplifies the decomposition.

Now, I started looking at this one when I started seeing failures on a
test that's added in a later PR. What's going on here is that, by doing
an upcast to a higher dtype and then cast down to integers, sometimes
there's an off-by-one error. I think this is fine, as the decomposition
is more accurate than the original function, which goes in line with
the whole PrimTorch effort.

cc ezyang mruberry ngimel Lezcano fdrocha

[ghstack-poisoned]

mruberry commented on torch/_refs/__init__.py:
```diff
@@ -4104,21 +4105,18 @@ def linspace(
     if py_any(isinstance(arg, complex) for arg in (start, end, steps)):
-        raise NotImplementedError
-    assert not isinstance(start, complex) and not isinstance(end, complex)  # for mypy
+        if dtype is None:
```
Why not use

https://github.com/pytorch/pytorch/blob/9f11ce7f67612d1c11f1a6a9b264779b27062e82/torch/_prims_common/__init__.py#L1006

?

lezcano replied:
I just reproduced the checks in core really. I'd vote to leave it as it is as here we'd need to perform casting between 4 different dtypes (start, end, step and dtype).

mruberry commented on torch/_refs/__init__.py:
```diff
@@ -4104,21 +4105,18 @@ def linspace(
-        if isinstance(end, FloatLike):
-            end = sym_int(end)
-
+    result_dtype = utils.type_to_dtype(utils.get_higher_type(type(start), type(end)))
```
Does this really not depend on steps?

mruberry commented on torch/_refs/__init__.py:
```diff
@@ -4135,41 +4133,25 @@ def linspace(
+        eps = step_size / 2
+        # torch.arange is a reduction, so we need precision here
+        rg = torch.arange(
+            0, 1 + eps, step_size, dtype=torch.float64, **factory_kwargs  # type: ignore[arg-type]
```
We should talk about this, too -- adding this `eps` here still seems wrong to me

lezcano replied:
IMO the eps is fine. Note that it's not a fixed eps, but a variable number that makes sure that everything just works.

lezcano replied:
That being said, if you have another proposal, we can discuss it.

mruberry commented on torch/_refs/__init__.py:
```diff
@@ -4135,41 +4133,25 @@ def linspace(
     elif start == end:
-        ret = torch.full((steps,), start, dtype=dtype, **factory_kwargs)  # type: ignore[call-overload]
+        return torch.full((steps,), start, dtype=dtype, **factory_kwargs)  # type: ignore[arg-type]
     else:
```
Nit: `else` isn't necessary here and would save an indent

mruberry commented on torch/_refs/__init__.py:
```diff
@@ -4135,41 +4133,25 @@ def linspace(
-            )
-
-    return ret
+        step_size = 1 / (steps - 1)
```
Add a comment describing the algorithm

mruberry commented on torch/_refs/__init__.py:
```diff
@@ -4135,41 +4133,25 @@ def linspace(
+        rg = torch.arange(
+            0, 1 + eps, step_size, dtype=torch.float64, **factory_kwargs  # type: ignore[arg-type]
+        )
+        float_dtype = (
```
Not really a `float_dtype` if it's complexdouble

lezcano committed:
Update on "Simplify linspace decomp and increase its tolerance"


This is an interesting one

Since this is an operation that's intrinsically defined on the reals,
we should perform the ops on that dtype always, and just cast to
the desired dtype at the end. This simplifies the decomposition.

Now, I started looking at this one when I started seeing failures on a
test that's added in a later PR. What's going on here is that, by doing
an upcast to a higher dtype and then cast down to integers, sometimes
there's an off-by-one error. I think this is fine, as the decomposition
is more accurate than the original function, which goes in line with
the whole PrimTorch effort.

cc ezyang mruberry ngimel Lezcano fdrocha

[ghstack-poisoned]

lezcano committed:
Update on "Simplify linspace decomp and increase its tolerance"


This is an interesting one

Since this is an operation that's intrinsically defined on the reals,
we should perform the ops on that dtype always, and just cast to
the desired dtype at the end. This simplifies the decomposition.

Now, I started looking at this one when I started seeing failures on a
test that's added in a later PR. What's going on here is that, by doing
an upcast to a higher dtype and then cast down to integers, sometimes
there's an off-by-one error. I think this is fine, as the decomposition
is more accurate than the original function, which goes in line with
the whole PrimTorch effort.

cc ezyang mruberry ngimel Lezcano fdrocha

[ghstack-poisoned]

lezcano committed:
Update on "Simplify linspace decomp and increase its tolerance"


This is an interesting one

Since this is an operation that's intrinsically defined on the reals,
we should perform the ops on that dtype always, and just cast to
the desired dtype at the end. This simplifies the decomposition.

Now, I started looking at this one when I started seeing failures on a
test that's added in a later PR. What's going on here is that, by doing
an upcast to a higher dtype and then cast down to integers, sometimes
there's an off-by-one error. I think this is fine, as the decomposition
is more accurate than the original function, which goes in line with
the whole PrimTorch effort.

cc ezyang mruberry ngimel Lezcano fdrocha

[ghstack-poisoned]

mruberry reviewed (approved):
Stamped

lezcano committed:
Update on "Simplify linspace decomp and increase its tolerance"


This is an interesting one

Since this is an operation that's intrinsically defined on the reals,
we should perform the ops on that dtype always, and just cast to
the desired dtype at the end. This simplifies the decomposition.

Now, I started looking at this one when I started seeing failures on a
test that's added in a later PR. What's going on here is that, by doing
an upcast to a higher dtype and then cast down to integers, sometimes
there's an off-by-one error. I think this is fine, as the decomposition
is more accurate than the original function, which goes in line with
the whole PrimTorch effort.

cc ezyang mruberry ngimel Lezcano fdrocha

[ghstack-poisoned]
